Problem Statement: Build a recommender system by using cosine simillarties score.

In [37]:
import pandas as pd
import numpy as np 
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine
import warnings
warnings.filterwarnings("ignore")

In [2]:
books = pd.read_csv("G:/data sceince/Assignments/Recommendation Systems/book (1).csv")

In [3]:
books.head()

Unnamed: 0  User.ID                                         Book.Title  \
0           1   276726                                Classical Mythology   
1           2   276729                                       Clara Callan   
2           3   276729                               Decision in Normandy   
3           4   276736  Flu: The Story of the Great Influenza Pandemic...   
4           5   276737                             The Mummies of Urumchi   

   Book.Rating  
0            5  
1            3  
2            6  
3            8  
4            6

In [12]:
books_df = books.rename({'User.ID':'userid','Book.Title':'booktitle','Book.Rating':'bookrating'},axis = 1)

In [13]:
books_df

Unnamed: 0  userid                                          booktitle  \
0              1  276726                                Classical Mythology   
1              2  276729                                       Clara Callan   
2              3  276729                               Decision in Normandy   
3              4  276736  Flu: The Story of the Great Influenza Pandemic...   
4              5  276737                             The Mummies of Urumchi   
...          ...     ...                                                ...   
9995        9996  162121       American Fried: Adventures of a Happy Eater.   
9996        9997  162121                              Cannibal In Manhattan   
9997        9998  162121                    How to Flirt: A Practical Guide   
9998        9999  162121                                           Twilight   
9999       10000  162129                       Kids Say the Darndest Things   

      bookrating  
0              5  
1              3  
2              6  
3              8  
4              6  
...          ...  
9995           7  
9996           9  
9997           7  
9998           8  
9999           6  

[10000 rows x 4 columns]

In [14]:
books1 = books_df.drop(['Unnamed: 0'], axis = 1)

In [15]:
books1

userid                                          booktitle  bookrating
0     276726                                Classical Mythology           5
1     276729                                       Clara Callan           3
2     276729                               Decision in Normandy           6
3     276736  Flu: The Story of the Great Influenza Pandemic...           8
4     276737                             The Mummies of Urumchi           6
...      ...                                                ...         ...
9995  162121       American Fried: Adventures of a Happy Eater.           7
9996  162121                              Cannibal In Manhattan           9
9997  162121                    How to Flirt: A Practical Guide           7
9998  162121                                           Twilight           8
9999  162129                       Kids Say the Darndest Things           6

[10000 rows x 3 columns]

In [ ]:
# extracting unique values from userid column, also getting the count of unique ids

In [16]:
len(books1['userid'].unique())

2182

In [26]:
array_user = books1['userid'].unique()

In [ ]:
# extracting unique values from Book.Title column, also getting the count of unique books

In [17]:
len(books1['booktitle'].unique())

9659

In [ ]:
# Creating Collaborative matrix using PIVOT_Table function

In [29]:
books_df1 = books1.pivot_table(index = 'userid',
                        columns = 'booktitle',
                        values = 'bookrating').reset_index(drop = True)

In [30]:
books_df1

booktitle   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
0                           NaN                                         NaN   
1                           NaN                                         NaN   
2                           NaN                                         NaN   
3                           NaN                                         NaN   
4                           NaN                                         NaN   
...                         ...                                         ...   
2177                        NaN                                         NaN   
2178                        NaN                                         NaN   
2179                        NaN                                         NaN   
2180                        NaN                                         NaN   
2181                        NaN                                         NaN   

booktitle   Repairing PC Drives &amp  '48  \
0                                NaN  NaN   
1                                NaN  NaN   
2                                NaN  NaN   
3                                NaN  NaN   
4                                NaN  NaN   
...                              ...  ...   
2177                             NaN  NaN   
2178                             NaN  NaN   
2179                             NaN  NaN   
2180                             NaN  NaN   
2181                             NaN  NaN   

booktitle  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
0                                                        NaN                       
1                                                        NaN                       
2                                                        NaN                       
3                                                        NaN                       
4                                                        NaN                       
...                                                      ...                       
2177                                                     NaN                       
2178                                                     NaN                       
2179                                                     NaN                       
2180                                                     NaN                       
2181                                                     NaN                       

booktitle  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
0                                                        NaN              
1                                                        NaN              
2                                                        NaN              
3                                                        NaN              
4                                                        NaN              
...                                                      ...              
2177                                                     NaN              
2178                                                     NaN              
2179                                                     NaN              
2180                                                     NaN              
2181                                                     NaN              

booktitle  01-01-00: A Novel of the Millennium  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   
...                                        ...   
2177                                       NaN   
2178                                       NaN   
2179                                       NaN   
2180                                       NaN   
2181                                       NaN   


In [ ]:
# Mapping the unique userids with the index  

In [31]:
books_df1.index = books1.userid.unique()

In [32]:
books_df1

booktitle   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
276726                      NaN                                         NaN   
276729                      NaN                                         NaN   
276736                      NaN                                         NaN   
276737                      NaN                                         NaN   
276744                      NaN                                         NaN   
...                         ...                                         ...   
162107                      NaN                                         NaN   
162109                      NaN                                         NaN   
162113                      NaN                                         NaN   
162121                      NaN                                         NaN   
162129                      NaN                                         NaN   

booktitle   Repairing PC Drives &amp  '48  \
276726                           NaN  NaN   
276729                           NaN  NaN   
276736                           NaN  NaN   
276737                           NaN  NaN   
276744                           NaN  NaN   
...                              ...  ...   
162107                           NaN  NaN   
162109                           NaN  NaN   
162113                           NaN  NaN   
162121                           NaN  NaN   
162129                           NaN  NaN   

booktitle  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
276726                                                   NaN                       
276729                                                   NaN                       
276736                                                   NaN                       
276737                                                   NaN                       
276744                                                   NaN                       
...                                                      ...                       
162107                                                   NaN                       
162109                                                   NaN                       
162113                                                   NaN                       
162121                                                   NaN                       
162129                                                   NaN                       

booktitle  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
276726                                                   NaN              
276729                                                   NaN              
276736                                                   NaN              
276737                                                   NaN              
276744                                                   NaN              
...                                                      ...              
162107                                                   NaN              
162109                                                   NaN              
162113                                                   NaN              
162121                                                   NaN              
162129                                                   NaN              

booktitle  01-01-00: A Novel of the Millennium  \
276726                                     NaN   
276729                                     NaN   
276736                                     NaN   
276737                                     NaN   
276744                                     NaN   
...                                        ...   
162107                                     NaN   
162109                                     NaN   
162113                                     NaN   
162121                                     NaN   
162129                                     NaN   


In [ ]:
# As we are required to calculate the similarities so replacing all the null values with 0.0

In [34]:
books_df1.fillna(0, inplace = True)

In [35]:
books_df1

booktitle   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
276726                      0.0                                         0.0   
276729                      0.0                                         0.0   
276736                      0.0                                         0.0   
276737                      0.0                                         0.0   
276744                      0.0                                         0.0   
...                         ...                                         ...   
162107                      0.0                                         0.0   
162109                      0.0                                         0.0   
162113                      0.0                                         0.0   
162121                      0.0                                         0.0   
162129                      0.0                                         0.0   

booktitle   Repairing PC Drives &amp  '48  \
276726                           0.0  0.0   
276729                           0.0  0.0   
276736                           0.0  0.0   
276737                           0.0  0.0   
276744                           0.0  0.0   
...                              ...  ...   
162107                           0.0  0.0   
162109                           0.0  0.0   
162113                           0.0  0.0   
162121                           0.0  0.0   
162129                           0.0  0.0   

booktitle  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
276726                                                   0.0                       
276729                                                   0.0                       
276736                                                   0.0                       
276737                                                   0.0                       
276744                                                   0.0                       
...                                                      ...                       
162107                                                   0.0                       
162109                                                   0.0                       
162113                                                   0.0                       
162121                                                   0.0                       
162129                                                   0.0                       

booktitle  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
276726                                                   0.0              
276729                                                   0.0              
276736                                                   0.0              
276737                                                   0.0              
276744                                                   0.0              
...                                                      ...              
162107                                                   0.0              
162109                                                   0.0              
162113                                                   0.0              
162121                                                   0.0              
162129                                                   0.0              

booktitle  01-01-00: A Novel of the Millennium  \
276726                                     0.0   
276729                                     0.0   
276736                                     0.0   
276737                                     0.0   
276744                                     0.0   
...                                        ...   
162107                                     0.0   
162109                                     0.0   
162113                                     0.0   
162121                                     0.0   
162129                                     0.0   


In [ ]:
# Calculating similarties using cosine metrics

In [102]:
user_sim = 1 - pairwise_distances(books_df1.values, metric = 'cosine')

In [103]:
user_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [104]:
user_sim_df = pd.DataFrame(user_sim)            # Converting the calculated similarities in dataframe
user_sim_df.iloc[0:5,0:5]                       # Displaying first 5 records

0    1    2    3    4
0  1.0  0.0  0.0  0.0  0.0
1  0.0  1.0  0.0  0.0  0.0
2  0.0  0.0  1.0  0.0  0.0
3  0.0  0.0  0.0  1.0  0.0
4  0.0  0.0  0.0  0.0  1.0

In [105]:
user_sim_df.index = books1.userid.unique()          # Mapping the index with original userids 
user_sim_df.columns = books1.userid.unique()        # Mapping the column_names with original userids

In [106]:
user_sim_df.iloc[0:5,0:5]

276726  276729  276736  276737  276744
276726     1.0     0.0     0.0     0.0     0.0
276729     0.0     1.0     0.0     0.0     0.0
276736     0.0     0.0     1.0     0.0     0.0
276737     0.0     0.0     0.0     1.0     0.0
276744     0.0     0.0     0.0     0.0     1.0

In [107]:
# As the similarities between the same userids will be always one so as we are required to see the similarities between different users so filling the diagonal values by 0

np.fill_diagonal(user_sim,0)         

In [108]:
user_sim_df.iloc[0:5,0:5]

276726  276729  276736  276737  276744
276726     0.0     0.0     0.0     0.0     0.0
276729     0.0     0.0     0.0     0.0     0.0
276736     0.0     0.0     0.0     0.0     0.0
276737     0.0     0.0     0.0     0.0     0.0
276744     0.0     0.0     0.0     0.0     0.0

In [109]:
# Finding out which users are having maximum simlarities between them

user_sim_df.idxmax(axis = 1)

276726    276726
276729    276726
276736    276726
276737    276726
276744    276726
           ...  
162107    276726
162109    276726
162113    161453
162121    276726
162129    276726
Length: 2182, dtype: int64

Inference : User 276736 and user 276726 having maximun similarities between them, user 162107 and user 276726 having maximum similaritie between them and so on..

In [82]:
# Displaying records of the users having similarities

books1[(books1['userid'] == 276744) | (books1['userid'] == 276726)]

userid               booktitle  bookrating
0  276726     Classical Mythology           5
5  276744  The Kitchen God's Wife           7

In [89]:
user_1 = books1[books1['userid'] == 162129]

In [90]:
user_2 = books1[books1['userid'] == 276726]

In [91]:
pd.merge(user_1,user_2, on = 'booktitle', how = 'outer')

userid_x                     booktitle  bookrating_x  userid_y  \
0  162129.0  Kids Say the Darndest Things           6.0       NaN   
1       NaN           Classical Mythology           NaN  276726.0   

   bookrating_y  
0           NaN  
1           5.0